### DISCUSSION TITLE
Average embeddings using SentenceEmbeddings

### DISCUSSION THREAD

**Discussion by pdurham2** [JohnSnowLabs/spark-nlp#5974] [Q&A] [answered]
Hello,

I am trying to create document level embeddings using the 100 dimension glove model and am having issues with the SetentenceEmbeddings annotator. My understanding is that by setting the inputCols to ["document", "embeddings"], the embeddings should then be averaged across the whole document. What I am finding is that the output vector seems to be concatenating the vectors for each sentence. In the example below, I provide three sentences and so I expect a single vector of 100 dimensions that is the average of the embeddings across all 3 sentences, but instead, I get a vector of length 300. If I provide 4 sentences, I get a length of 400, etc. Am I defining my pipeline incorrectly to produce the document-level embeddings? I am using sparknlp 3.2.1 and Spark 3.1.0.

Thank you in advance for any help!

```
import sparknlp
from sparknlp.base import *
from sparknlp.annotator import *
from pyspark.ml import Pipeline
import pyspark.sql.functions as F

document_assembler = DocumentAssembler() \
.setInputCol("text") \
.setOutputCol("document") \
.setIdCol("id") \
.setCleanupMode("shrink") # removes new lines, tabs, and merges multiple blank spaces into one space 

sentencer = SentenceDetector()\
  .setInputCols(["document"]) \
  .setOutputCol("sentences")

tokenizer = Tokenizer() \
.setInputCols(["sentences"]) \
.setOutputCol("token") \
.setSplitChars(['-']) \
.setContextChars(['(', ')', '?', '!', '#', '@'])

embeddings = WordEmbeddingsModel.pretrained("glove_100d", "en") \
      .setInputCols("sentences", "token") \
      .setOutputCol("embeddings")

embeddingsDocument = SentenceEmbeddings() \
    .setInputCols(["document", "embeddings"]) \
    .setOutputCol("document_embeddings") \
    .setPoolingStrategy("AVERAGE")

embeddingsFinisher = EmbeddingsFinisher() \
    .setInputCols(["document_embeddings"]) \
    .setOutputCols("finished_embeddings") \
    .setOutputAsVector(True) \
    .setCleanAnnotations(False)
      

nlp_pipeline = Pipeline(stages=[document_assembler, sentencer, tokenizer, embeddings, embeddingsDocument, embeddingsFinisher])

# example query
t = 'this is the first sentence. here is another one. this is the last one.'
pipeline_model.transform(spark.createDataFrame([[0, t]], ["id", "text"])).show(truncate=False)
```

![image](https://user-images.githubusercontent.com/17155167/130458327-3833e31d-9c66-4b7f-bc97-a9df259744db.png)

**Comment by maziyarpanahi** [comment, marked as answer]
Hi,

Since you are using `WordEmbeddingsModel`, and for this annotator (`GloVe`) there is no maximum sequence unlike transformers (BERT, ALBERT, RoBERTa, etc.), you can just remove SentenceDetector, use `document` for setInputCols in all the annotators that come after DocumentAssembler. 

Could you please try that and see what happens?

UPDATE: This should be fixed in our new release: https://github.com/JohnSnowLabs/spark-nlp/releases/tag/3.2.3

**Reply by pdurham2** [reply]
Removing the SentenceDector and using document as the input for the embeddings produces a single vector with the correct 100 dimensions. Thank you for the help!

**Reply by maziyarpanahi** [reply]
You are very welcome. It is a best practice to either keep everything at the document level or sentence throughout the pipeline to have a unified entry point to be document level or sentence level.

**Reply by pdurham2** [reply]
That makes sense, if I wanted to instead create sentence level embeddings, should I repleace all mentions of 'document' with 'sentences' after the SentenceDetector in the pipeline as is below?

```
import sparknlp
from sparknlp.base import *
from sparknlp.annotator import *
from pyspark.ml import Pipeline
import pyspark.sql.functions as F

document_assembler = DocumentAssembler() \
.setInputCol("text") \
.setOutputCol("document") \
.setIdCol("id") \
.setCleanupMode("shrink") # removes new lines, tabs, and merges multiple blank spaces into one space 

sentencer = SentenceDetector()\
  .setInputCols(["document"]) \
  .setOutputCol("sentences")

tokenizer = Tokenizer() \
.setInputCols(["sentences"]) \
.setOutputCol("token") \
.setSplitChars(['-']) \
.setContextChars(['(', ')', '?', '!', '#', '@'])

embeddings = WordEmbeddingsModel.pretrained("glove_100d", "en") \
      .setInputCols("sentences", "token") \
      .setOutputCol("embeddings")

embeddingsSentence = SentenceEmbeddings() \
    .setInputCols(["sentences", "embeddings"]) \
    .setOutputCol("sentence_embeddings") \
    .setPoolingStrategy("AVERAGE")

embeddingsFinisher = EmbeddingsFinisher() \
    .setInputCols(["sentence_embeddings"]) \
    .setOutputCols("finished_embeddings") \
    .setOutputAsVector(True) \
    .setCleanAnnotations(False)
      

nlp_pipeline = Pipeline(stages=[document_assembler, sentencer, tokenizer, embeddings, embeddingsSentence, embeddingsFinisher])
```

Using this pipeline, I am seeing the same problem where SentenceEmbeddings returns 2 vectors of 200 dimensions rather than 2 100 dimension vectors. I confirmed that the length of the actual embeddings is 100. Is there something incorrect in my definitions of the annotators?

```
sample = pipeline_model.transform(
    spark.createDataFrame([[0, 'this is a test. and also this is a much longer sentence with lots of words.']], ["id", "text"])
).withColumn("finished_embeddings", F.explode("finished_embeddings"))
```
![image](https://user-images.githubusercontent.com/17155167/130518037-43cb6a16-8134-4675-9676-f299793384f8.png)

**Reply by maziyarpanahi** [reply]
The result of the column `sentence_embeddings` must be vectors in 100 dimensions with the number of sentences. (if you have 4 sentences, you must have an array with 4 items and each must be 100 Floats. The dimension will never be something more or less than the dimension of the original word embeddings, since it just averages them and not contacting)

**Reply by pdurham2** [reply]
That is what I expected to see as well, but the `sentence_embeddings` is returning three arrays with length 300. If I change the input to be four sentences, I get four arrays with length 400 instead. Is it possible that the function that is averaging the arrays is actually concatenating the arrays when it performs the summation during the average?

![image](https://user-images.githubusercontent.com/17155167/130646927-e658bfc4-911b-4941-ab58-bb292097525e.png)

**Reply by maziyarpanahi** [reply]
I will debug this with your example and see what happens. I think something went wrong in your calculations of dimension but if it’s the bug will fix it in the next release

**Reply by maziyarpanahi** [reply]
In the first test, this doesn't happen in Scala: 

```Scala
pipelineDF.select(size(pipelineDF("sentence_embeddings.embeddings").getItem(0)).as("sentence_embeddings_size")).show
```

This will give the correct dimension. I am not sure if it's the `F.size`/`F.col` not correctly work or it's in the Python side. (which doesn't make sense since the Python side is actually the code coming from Scala). 

I will dig more to see why there is a difference. There should be no need for explode since we take the item from the array and check the size.

**Reply by maziyarpanahi** [reply]
@pdurham2 This should be fixed in our new release: https://github.com/JohnSnowLabs/spark-nlp/releases/tag/3.2.3

**Reply by pdurham2** [reply]
@maziyarpanahi Great, thank you!
